### ISSUE TITLE
Lint/UselessSetterCall + at_exit = False Positive

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#14481]
## Expected behavior

No warning, because the setter is in fact not useless as it stands.

## Actual behavior

```Useless setter call to local variable `foo`. [[Correctable] Lint/UselessSetterCall]```

## Steps to reproduce the problem

```ruby
# Foo
class Foo < Hash
  def initialize
    super
    at_exit { save }
  end

  def save
    File.write('/tmp/save')
  end
end

def main
  foo = Foo.new
  foo['bar'] = 'baz
  # foo.save # not really needed since we have a catch-all auto-save
end

main
```

## RuboCop version

```
ruby      2025-08-01 30a20bc16
rubocop   2025-08-12 c34098ee0
System Version: macOS 15.6 (24G84)
Model Identifier: MacBookAir10,1
```

**Comment by koic** [comment]
This looks like the expected behavior. As shown in the documentation:
https://docs.rubocop.org/rubocop/1.80/cops_lint.html#lintuselesssettercall

The issue is resolved by explicitly returning the receiver from `main`:

```diff
  class Foo < Hash
    def initialize
      super
      at_exit { puts self }
    end
  end

  def main
    foo = Foo.new
    foo['bar'] = 'baz'
+   foo
  end

  main
```

It can also be resolved with `rubocop -A`.

**Comment by forthrin** [comment]
True, but this isn't strictly needed by the program itself. By leveraging `at_exit` (sort of an emergency destructor, of which Ruby has none) the object is guaranteed to be saved whether the program succeeds or raises an exception. Thus there is no need for an explicit save.

Is analyzing the code for this circumstance complicated?

And if there is a less sledgehammer way to prevent data loss on exception, very interested I am.

**Comment by Earlopain** [comment]
For your example, just use `ensure` and call save there, that way you have code flow that isn't so confusing with `at_exit`

```rb
def main
  foo = Foo.new
  foo['bar'] = 'baz'
ensure
  foo.save
end
```

**Comment by forthrin** [comment]
This is just a minified example. The code will obviously have lots of writes to the foo object around and about.

You are however saluted for the `ensure` trick. Didn't know you could do that independent of a `rescue`. So the `ensure` can be added to the constructor. But alas it doesn't seem to trigger on an arbitrary exception, and then data in memory is lost.

- Program end: YES
- Explicit exit: YES
- Exception: NO

**Comment by Earlopain** [comment]
Well, it obviously wont run when there is no exception happening in the constructor. Anyways, I don't think there is really much of an issue with rubocop here. The offense may be a bit redundant with your example but it doesn't provide wrong autocorrect. RuboCop is not able to fix this since `at_exit` introduces code-flow that it can't know about. Just by looking at the code with the offense (which is all that rubocop does really) it doesn't see your `at_exit` (which can also be a different file) at some other place.

BTW, using `at_exit` like this means that these objects are never garbage collected, even if there are otherwise no references to them anymore.

**Comment by forthrin** [comment]
Is this a hole in the Ruby memory deallocation scheme that should be addressed? At any rate, the operating system will surely free all memory allocated by a given process when it goes to meet its maker.

I was quite expecting someone to deem understanding this type of eventuality to be out of the parser's scope, but was interested in presenting it to see what kind of aspects that would come up.

**Comment by Earlopain** [comment]
You are holding a reference to the object in the `at_exit` hook, you are saying to ruby you always need this object until the very end (how could you save it if ruby took it away before then?). Of course it will be freed after but for long-running processes this can quickly become a problem. It is basically a [finalizer](https://docs.ruby-lang.org/en/master/ObjectSpace.html#method-c-define_finalizer) that references the object it is supposed to finalize. For that ruby would emit a warning (you can't use warning-free finalizers for what you want to do).

**Comment by koic** [comment]
In any case, since this is not a false positive related to `Lint/UselessSetterCall`. This will be closed as an issue.

**Comment by forthrin** [comment]
Concurring. The underlying issue seems more to be how to avoid data loss. It seems the recommended practice is actually to save data proactively (at the cost of overhead), rather than trying to hook into thinkable exceptions and signals.
